File: Day12/rainRisk.py
# part 1
with open ('input.txt') as inp:
    EW = 0
    NS = 0
    dir = 90

    for line in inp:
        action = line[0]
        num = int(line[1:])
        if action == "F" and dir == 90 or action == "E":
            EW += num
        elif action == "F" and dir == 180 or action == "S":
            NS -= num
        elif action == "F" and dir == 270 or action == "W":
            EW -= num
        elif action == "F" and dir == 0 or action == "N":
            NS += num
        elif action == "L":
            dir -= num
            if dir == 360 or dir == -360:
                dir = 0
            if dir == -90:
                dir = 270
            if dir == -180 or dir == 540:
                dir = 180
            if dir == -270 or dir == 450:
                dir = 90

        elif action == "R":
            dir += num
            if dir == 360 or dir == -360:
                dir = 0
            if dir == -90:
                dir = 270
            if dir == -180 or dir == 540:
                dir = 180
            if dir == -270 or dir == 450:
                dir = 90

print(f"Manhatten Distance: {abs(NS)+ abs(EW)}")

# part 2
with open ('input.txt') as inp:
    ship_ew = 0
    ship_ns = 0
    way_ew = 10
    way_ns = 1

    for line in inp:
        action = line[0]
        num = int(line[1:]) 

        if action == "F":
            ship_ns += (way_ns * num)
            ship_ew += (way_ew * num) 
        elif action == "N":
            way_ns += num
        elif action == "S":
            way_ns -= num
        elif action == "E":
            way_ew += num
        elif action == "W":
            way_ew -= num

        elif action == "L" and num == 90 or action == "R" and num == 270:
            way_ns, way_ew = way_ew, -way_ns
        elif action == "L" and num == 270 or action == "R" and num ==90:
            way_ns, way_ew = -way_ew, way_ns
        elif action == "L" and num == 180 or action == "R" and num == 180:
            way_ns = -way_ns
            way_ew = -way_ew

print(f"Manhattan Distance: {abs(ship_ns)+ abs(ship_ew)}")

File: Day12/rainRisk.ts
const fs = require('fs')
const lines = fs.readFileSync('input.txt', {encoding: 'utf-8'}).split('\n')

// part 1
const compass = {
  0: "N",
  90: "E",
  180: "S",
  270: "W",
};
class Ship {
  protected _degrees: keyof typeof compass;
  protected _position: [x: number, y: number];

  public constructor() {
    this._degrees = 90;
    this._position = [0, 0];
  }

  public move(direction: string, value: number) {
    switch (direction) {
      case "N":
        this._position[1] += value;
        break;
      case "E":
        this._position[0] += value;
        break;
      case "S":
        this._position[1] -= value;
        break;
      case "W":
        this._position[0] -= value;
        break;
      case "F":
        this.move(compass[this._degrees], value);
        break;
      case "L":
        this._degrees -= value;
        if (this._degrees < 0) this._degrees += 360;
        break;
      case "R":
        this._degrees += value;
        if (this._degrees >= 360) this._degrees -= 360;
        break;
    }
  }

  public get position(): [x: number, y: number] {
    return [...this._position];
  }
}
const ship = new Ship();

lines.forEach((line) => {
  const action = line.substr(0, 1);
  const value = parseInt(line.substr(1));
  ship.move(action, value);
});

const [x, y] = ship.position;
console.log(Math.abs(x) + Math.abs(y));

// part 2
class BetterShip {
  protected _position: [x: number, y: number];
  protected _waypoint: [x: number, y: number];

  public constructor() {
    this._position = [0, 0];
    this._waypoint = [10, 1];
  }

  public move(direction: string, value: number) {
    switch (direction) {
      case "N":
        this._waypoint[1] += value;
        break;
      case "E":
        this._waypoint[0] += value;
        break;
      case "S":
        this._waypoint[1] -= value;
        break;
      case "W":
        this._waypoint[0] -= value;
        break;
      case "F":
        this._position[0] += this._waypoint[0] * value;
        this._position[1] += this._waypoint[1] * value;
        break;
      case "L":
        switch (value) {
          case 90:
            this._waypoint.reverse();
            this._waypoint[0] *= -1;
            break;
          case 180:
            this._waypoint[0] *= -1;
            this._waypoint[1] *= -1;
            break;
          case 270:
            this._waypoint.reverse();
            this._waypoint[1] *= -1;
            break;
        }
        break;
      case "R":
        switch (value) {
          case 90:
            this._waypoint.reverse();
            this._waypoint[1] *= -1;
            break;
          case 180:
            this._waypoint[0] *= -1;
            this._waypoint[1] *= -1;
            break;
          case 270:
            this._waypoint.reverse();
            this._waypoint[0] *= -1;
            break;
        }
        break;
    }
  }

  public get position(): [x: number, y: number] {
    return [...this._position];
  }
}
const betterShip = new BetterShip();

lines.forEach((line) => {
  const action = line.substr(0, 1);
  const value = parseInt(line.substr(1));
  betterShip.move(action, value);
});

// const [x, y] = betterShip.position;
// console.log(Math.abs(x) + Math.abs(y));